File: 김소희/연구소.java
import java.io.*;
import java.util.*;

public class Main {

    static int N, M;
    static int[][] map;
    static List<int[]> empties = new ArrayList<>();
    static List<int[]> virus = new ArrayList<>();
    static int answer = 0;

    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    // 바이러스 확산 BFS
    static int bfs() {
        boolean[][] visited = new boolean[N][M];
        Queue<int[]> q = new ArrayDeque<>();

        for (int[] v : virus) {
            q.offer(v);
            visited[v[0]][v[1]] = true;
        }

        int infected = 0;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int r = cur[0], c = cur[1];

            for (int i = 0; i < 4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];

                if (nr < 0 || nr >= N || nc < 0 || nc >= M) continue;

                if (!visited[nr][nc] && map[nr][nc] == 0) {
                    visited[nr][nc] = true;
                    infected++;
                    q.offer(new int[]{nr, nc});
                }
            }
        }
        return infected;
    }

    // 벽 3개 조합 DFS
    static void dfs(int idx, int wall) {

        if (wall == 3) {
            int infected = bfs();
            int safe = empties.size() - 3 - infected;
            answer = Math.max(answer, safe);
            return;
        }

        for (int i = idx; i < empties.size(); i++) {
            int r = empties.get(i)[0];
            int c = empties.get(i)[1];

            map[r][c] = 1;          // 벽 세우기
            dfs(i + 1, wall + 1);   // 다음 인덱스부터
            map[r][c] = 0;          // 복구
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        map = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
                if (map[i][j] == 0) empties.add(new int[]{i, j});
                else if (map[i][j] == 2) virus.add(new int[]{i, j});
            }
        }

        dfs(0, 0);
        System.out.println(answer);
    }
}

File: 김소희/연구소.py
from collections import deque

N, M = map(int, input().split())
Map2L = [list(map(int, input().split())) for _ in range(N)]
 
empties = []
virus = []

for r in range(N):
    for c in range(M):
        if Map2L[r][c] == 0:
            empties.append((r, c))
        elif Map2L[r][c] == 2:
            virus.append((r, c))

answer = 0

def bfs():
    visited = [[False]*M for _ in range(N)]
    Q = deque()

    for r, c in virus:
        Q.append((r, c))
        visited[r][c] = True

    cnt = 0
    while Q:
        r, c = Q.popleft()
        for dr, dc in [(-1,0),(1,0),(0,-1),(0,1)]:
            nr, nc = r+dr, c+dc
            if 0 <= nr < N and 0 <= nc < M:
                if not visited[nr][nc] and Map2L[nr][nc] == 0:
                    visited[nr][nc] = True
                    cnt += 1
                    Q.append((nr, nc))
    return cnt


def dfs(idx, wall):
    global answer
    if wall == 3:
        infected = bfs()
        answer = max(answer, len(empties) - 3 - infected)
        return

    for i in range(idx, len(empties)):
        r, c = empties[i]
        Map2L[r][c] = 1
        dfs(i+1, wall+1)
        Map2L[r][c] = 0


dfs(0, 0)
print(answer)
